compiler/read_file.cpp

```diff
@@ -0,0 +1,42 @@
+/*
+* Copyright: Copyright ssiatras (Stathis Siatras) 2019.
+* License:   $(LINK2 http://www.boost.org/LICENSE_1_0.txt, Boost License 1.0)
+* Authors:   ssiatras (Stathis Siatras)
+*/
+#include "read_file.h"
+#include <fstream>
+#include <sstream>
+#include <string>
+#include <cstring>
+
+/*
+ * This function reads the content of a text file as a whole into memory as a
+ * NULL-terminated, contiguous array of characters.
+ *
+ * Params:
+ *    filepath = The path of the file to be read.
+ * Returns:
+ *    A pointer to an array of characters (the content of the file).
+ */
+const char *read_file_into_memory(const char *filepath) {
+  if (filepath == NULL) {
+    return NULL;
+  }
+
+  std::ifstream finput(filepath, std::ios::in | std::ios::binary);
+  if (!finput) {
+  	return NULL;
+  }
+
+  std::ostringstream contents;
+  contents << finput.rdbuf();
+  finput.close();
+
+  std::string str = contents.str();
+  long int fsize = str.size();
+
+  char *cstr = new char[fsize + 1];
+  strcpy(cstr, str.c_str());
+
```

Comment:
You don't need to allocate memory yourself and copy the string as you already have it in `str`. You just have to return `str.c_str()`.

Reply:
However, you then want to _not_ `delete[]` the memory you're getting. Because it's now opaque how the memory is allocated so we don't want to free it (and we don't need to).

Reply:
c_str() returns an array of chars in stack. It won't exist after the exit of the function, that's why we have to dynamically allocate it.

Reply:
It could not possibly do that. Well, at least in any sane way. Let's say function `test()` calls `c_str()`. Then, `c_str()` would have to somehow have access to the stack of `test()` in order to allocate an array of chars that would still be valid when `c_str()` returns.

Reply:
I tried to return what c_str() returns. The object string from str() is destructed and the its internal memory is cleaned up. That way, the caller of our function receives a dangling pointer.

Reply:
Yes, this is a reasonable concern and it's C++ doing magic. And unfortunately, we have to add a little bit of magic ourselves to prevent the destruction.
Note that this is not a minor detail to just make a copy of the string. The input file might be large, i.e. large enough that 2 of them don't fit into memory.
So, you can create a dummy global variable `std::string` and save there the `contents.str()`. Because it is global, it won't be destroyed when the function exits. The other alternative is to return to your previous solution.
Both alternatives are almost equally bad, so you may want for us to try to think something better (you can do the same too).
For now I propose instead of returning `const char *`, return `std::string`. And just get the contents with `c_str()` in `main()`. It's a reasonable tradeoff.

Reply:
Alternative:
```cpp
size_t get_file_length(std::ifstream &is) {
  is.seekg(0, is.end);
  size_t length = is.tellg();
  is.seekg(0, is.beg);
  return length;
}

std::ifstream is (filepath, std::ifstream::binary);
assert(is);
const char *ret = new char [get_file_length(is)];
is.read(ret, length);
return ret;
```
This is sort of like your first solution but I think it's cleaner.
Pick whatever solution you like of the 2.

Reply:
Sorry, I missed that. If I have forgotten a comment / PR / etc. for a long time (say more than 3-4 days), ping me (by e.g. referencing me: @baziotis ). This is also true for other projects that you'll contribute. It's not the easiest thing in the world to keep up with _all_ the activity in a project so people forget stuff or miss them completely.

Now, to your question. No, we don't need to. You can still pass `const char *`.